README
先看原程序的Readme:
权责声名：
    本软件仅供研究学习之用，无意对实达的交换机及认证机制做任何抵触性行为。
    本软件不可用于任何商业和不良用途，否则责任自负。
    本软件不保证在任何环境下能够通过，本人也不保证能按时按网友要求改进本软件，其编写及维护纯属个人爱好，有可能随时被终止。
    本软件不保证经过严格测试对机器无害，由于未知的使用环境或不当的使用对计算机造成的损害，责任由使用者全部承担。


★★★由于任何不遵守上叙条例引起的纠纷，均与本人无关，如不同意该声名请不要使用该软件，谢谢合作。★★★


netxray@byhh 2004/9/7 15:20

mystar Ver0.1
================
ps: 本程序在redhat7.2/9.0 + libnet v1.1.1 + libpcap v0.8.3 + 沁园 + 2004.09下通过
代码写的很糟糕，不要笑偶 ^_^

***************************************************************************
csip amoblin@gmail.com

接下来是我添加和更改的:

aclocal.m4			aclocal
AUTHORS				配置要求
autom4te.cache/		文件夹
autoscan.log		autoscan命令生成
back_windows.sh		自己编写的转本工程下unix格式文本文档为dos文本文档的脚本
bin/				可执行文件所在目录
ChangeLog			开发日志
configure.in		autoscan命令生成的configure.scan改名后修改得到
COPYING
data				图形设计，配置文件
depcomp
dia					程序相关的源图
dia_png				程序相关的png图
doc					程序相关文档
GPL					GNU要求
install-sh
Makefile			自己写的Makefile文件
Makefile.am			自己写的，用于生成Makefile文件
missing
NEWS
obj					中间模块文件
pdf					程序相关pdf文档
README				本文档
ref					参考资料
shot				程序运行截图
src					程序源代码
makeall				生成程序源代码清单的bash脚本


xml库：xqilla, xerces-c
/usr/include/

required:
libxqilla-dev
libxerces-c2-dev
libboost-dev
libgtkmm-2.4-dev


ChangeLog
2009年05月26日 星期二 22时48分56秒
	完成uml图。

	以下是参考链接：
	make : http://www.chinalinuxpub.com/doc/pro/gmake.html
	libnet : http://blog.chinaunix.net/u2/82249/showart_1741727.html
	libpcap: http://www.diybl.com/course/3_program/c++/cppjs/2008324/106112.html
	tinyxml: http://www.wangchao.net.cn/bbsdetail_895762.html
			http://www.cnblogs.com/peterzb/articles/1456364.html
	glade 3 : http://www.gnome-cn.org:19580/gnomecn/Filescenter/documents/tutorial/glade-advance-tutorial
			http://www.cppblog.com/len/archive/2008/03/27/45548.html
	gtkmm:http://zh.wikipedia.org/wiki/Gtkmm
	libnotify:http://apt-blog.co.cc/archives/386.html
	dbus :http://www.cppblog.com/iuranus/archive/2009/05/13/69894.html
		http://www.ibm.com/developerworks/cn/linux/l-dbus.html

2009年05月24日 星期日 15时40分12秒
	修改论文，明确结构;
	增加src中各文件的说明注释，便于程序源代码的打印。
	去掉back_window.sh的后缀
	增加makeall脚本，将源文件统一于一个文件，以生成程序源代码清单
	修改pdf文件夹为doc_pdf，增加doc2pdf脚本
	增加dia的TODO
	确定翻译文章：lordan or eap white paper?
2009年05月13日 星期三 19时12分55秒
	完成论文封面。
2009年04月22日 星期三 01时22分53秒
	整理文档，有几个问题：
	1。MAC帧，802.1x协议，EAP协议，RADIUS协议，它们之间什么关系？
	2.帧的研究
2009年04月21日 星期二 12时57分37秒
	用libnm_glib实现了检测网口状态。
	14:23 实现了自动重连！
2009年04月20日 星期一 21时55分20秒
	完成notify功能。用libnotify-1.0
	修复了自动登录功能
	TASK:1.dbus 2.老师布置的任务
	22:47 现在完成了dbus的编译环境，api还没有找到。。。
	23:35 完成文本文档由unix到dos格式的批量转换，便于老师在Windows下阅读。
2009年04月20日 星期一 14时47分08秒
	今天下午见李老师，向老师详细报告了sendpacket的内容，并把整个工程的结构向老师描述了一遍。
	布置如下任务：
	RADIUS和EAPOL的联系,RADIUS用什么协议？
	认证服务器的MAC地址交换机如何得到？
	服务器对激活包的返回包是什么？
	Alog和MD5的了解
	三次握手
2009年04月20日 星期一 00时57分08秒
	研读了sendpacket文件，搞定！下一步是blog和md5文件的研究。还有dbus的实现...
2009年04月10日 星期五 08时08分25秒
	dropbox的Public不能下载了，这下没地方发布了，想来想去只有google code了。于是在google code上创建项目，重新用上了svn.
2009年04月09日 星期四 14时16分28秒
	用recordmydesktop录制了使用效果演示，默认是ogv格式的，又用mencoder转为avi格式的。
	提交更新错误，难道因为好久没提交了，删错文件了？索性将版本控制由svn改为git了，虽然对svn还没有完全了解，而且对git更不了解。。。
2009年04月09日 星期四 13时35分41秒
	完成tray图标的修改，初始时有一红杠，表示不连通，当认证成功后，红杠消失，可以上网。
2009年04月09日 星期四 13时08分53秒
	好几天没写了，一来在忙复试，二来在研究gmChat。现在gMstar已经差不多了，作了tray图标，感觉不好看，又更新了一个，很像NetWorkManager的。传到了dropbox上，在校内贴了url，期待大家的测试！
	看上次的任务，第3条完成了
2009年04月03日 星期五 01时07分39秒
	今天收获很大！特别感谢lerosua前辈的热心指导！是lerosua让我学会了使用线程，还有参考了lerosua的大作gmlive，学会了statusicon的使用！现在锐捷可以最小化到托盘了，并且我经过查资料和实践，发现gtkmm要用glade的话，最好不要继承自Window，除非你只想使用glade文件的一部分，并且要自己再建一遍ui。网上有继承自Main的，而我在gMystar里使用的不继承。用一个Window指针做成员，可以直接指向初始化的glade界面！其实本来gMystar一直是继承自Mystar的，但由于信号的连接要静态成员函数，而Mystar的authen一静态几乎全部都静态了，所以我决定让Mystar做gMystar的成员。
	现在可以发布使用了，剩下的是优化了！
	下面的任务：1。信号的断开2.程序的退出3。打包发布4.交叉编译到windows5.基础学习
2009年04月01日 星期三 15时18分49秒
	今天下载了chat聊天软件，作为新工程的开始。而gmystar还没有一点进展：一直在研究进程，线程！
	下一步要学习的任务：1.c++学习，包括const，static等的理解 2.进程学习 3.线程学习 4.信号处理
2009年03月31日 星期二 21时57分00秒
	vim重新用上了时间戳，以后ChangeLog不用自己输时间了！
2009-03-31 21:09
	今晚研究tray。把main从gmystar.cc中分离出来了，提高gmystar的可重用性.
2009-03-31 17:30
	成功实现xml配置文件的解析！又修改Makefile文件，实现后缀法批量编译，学习了静态连接和动态连接，makefile函数。
	下面的任务：1。glade设计2.信号处理
2009-03-30 23:51
	今天会用string类了，get_text搞定。晚上研究xml到现在仍没成功。解析器太多，不知如何选择。有xerces-c,libxml,libxml2,libxml++,expat,xqilla,libpath,tinyxml,tinyxml++等。。。决定用tinyxml
	那么得研究gcc和g++的参数问题了，具体是编译过程额研究，什么动态库，静态库，include，lib，预处理，汇编，链接等。明天把这一块搞明白，希望可以解析xml.
2009-03-27 00:17
	今晚（其实是26号晚）准备研究那个大循环，想拆成分别发包，但好像现在时机还不成熟，我对包的机制还不太了解。。。
	UI中除了读取配置文件显示外，还要读取最终用户输入或修改的数据，set_text的参数可以是char*，但get_text()的返回值是GLib::ustring，是utf8 string，不知道如何转换成char*？
2009-03-25 11:32
	今天上午完成了svn+apache2的配置.现在开始用svn进行版本控制了，只是还不会用分支。。。
	那么下面要做的是：1.信号处理;2.glade设计;3.打包处理
2009-03-25 01:22
	才发现每次写都过了十二点。。。
	今晚（准确说是昨晚）小研究了一下编码转换，这次终于成功了！
	学习了中文gb2312的编码，utf8编码太复杂了，留待以后进一步的学习，现在发布是没问题了。
	今天安装了git，那么现在cvs，git，svn全有了，还有一个教barra什么的，改天小觑一下。。
	打包看了下，得从autotool做起，太复杂，最后做吧。
	信号处理得好好研究一下，还有glade和gtkmm的学习！
	那么下面要做的是：1.信号处理;2.glade设计;3.cvs搭建4.打包处理
2009-03-24
	昨晚开始研究sendpackage.cc文件，原来没有设置ip，所以老出现ip绑定错误;没有读取配置文件种的echointerval，所以是假的keep sending echo。
	编码问题，暂时不会解决，不过我现在把十六进制码输出出来了，有空可以细细研究一下。
	现在已可以用gMystar上网了，只是信号处理部分还没整明白，ctrl+c还不能logout，而是中断程序，网还联着。。。但不send echo了，那我想也持续不了多久吧？
	还有GUI，网能连上后就假死，真无语。
	修改了配置文件，增加了install，复制gMystar,gMystar.glade,gMystar.conf到~/bin
	下面要做的是：1.编码转换;2.信号处理;3.glade设计;4.打包处理5.cvs搭建
2009-03-14
	做了一晚上的编码转换，还是还有搞定!iconv的C函数介绍太少了，iconv太难用了。。。现在能把gbk转换成乱七八糟的东西，也能转换成空，就是不能正确转换！
	研究了一下EAP，将原来的search_server方法更名为eapol，这样更符合逻辑，因为其中包含了整个报文的收发，而不仅仅是寻找服务器。这个函数和那张图现在高度对应了！
	下一步继续编码转换，还有就是研究sendpackage.cc文件了！
	今晚初次安装Debian，现在还在配置apt，之前的界面都很不错，安装也很快，就这里慢！难道是联网的缘故？
	对了，还有问题，发的包可能不对，因为老显示IP绑定错误，而其实应是达到用户数量上限，即别人已用了。这在研究sendpackage时能找到答案吧！
	现在是00:49,还在配置apt，索性把网线拔了，应该会快点吧，装完就睡！
2009-03-08
	用gdb和插入printf信息来调试search_server函数，简化为一次验证，并且运行成功
	iconv改来自服务器的gbk编码信息未果，iconv_open和iconv_close都会用了，但iconv函数的size_t *参数不好填，用int *代替还不行！明天继续查！
	界面还有改：GUI中默认回车就可联，但初始光标还应在最上面的输入框里，而focus到连接按钮上；命令行下通过参数控制GUI的显示，类似mplayer那样，用--nogui参数不显示GUI,但这样的话命令行就要自动连接了！GuI里面也可以设自动连接的其实！
	说干就干！现在23:58...
	现在00:37！现在有GUI和CLI两种模式，Makefile里也有了run和runc来运行gui和cli，原来c++的类实例化时，若无参则连括号都不要！
	在CLI模式下添加help没有成功，关于静态函数的用法明天查书吧！睡觉！已00:47了！
2009-02-08
	找到KStar，C++和qt改进的mystar。
	用dia画了search_server的流程图，
2009-01-24
	TODO:notification area show
2009-01-23
	TODO:xmldom解析mystar.xml,用xerces-c库和xqilla库
2009-01-21
	对象化大致完成。可运行。
	有User类，MD5类，Mystar类，gMystar类。
	Makefile文件也实现了分目录
2009-01-20
	学会了makefile多文件部署。之前已将mystar.cc构造成功。
	模块化myerr.o, sendpaket.o,User.o, MD5.o, Mystar.o,gMystar.o
	初步学会用glade进行界面设计，能够连接gtkmm和libglade进行事件关联。
	撤销了静态链接
2008-11-25
	改mystar.c 为mystar.cc，重构了代码，建了mystar类。但不能运行。
2008-11-23
	cvs配置完成，这是第一份在cvs下的添加，成功。
2008-11-14
	格式化工作：用indent格式化为gnu风格，又把那些十六进制数排的更好看些。
	修改为自动退出，当认证成功或失败时。因为服务器用的gb2312编码，而我用utf8，所以显示的信息是乱码，为了能转换编码，只好退出从而在脚本中用iconv转换。
	启用fakeAddress，虽然现在还不知道如何使用。
	修改目录，使更美观。用bin存可执行程序，obj下放中间文件，本来要把.c和.h放入src或lib中，但makefile貌似最好放在首目录下，所以就没有这样做了。这样也好，直接编辑，直接编译，直接运行。
	当然要修改makefile文件了。再加一个run选项！但没成功。


src/MD5.cc
/* MD5.c - an implementation of md5 algorithm */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

#include "MD5.h"

/* Constants for MD5Transform routine. */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

MD5::MD5()
{
}

/* MD5 initialization. Begins an MD5 operation, writing a new context.
 */
void MD5::MD5Init (MD5_CTX * context)
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants.
*/
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
void MD5::MD5Update (MD5_CTX *context, unsigned char *input, unsigned int inputLen)
{
  unsigned int i, index, partLen;

  /* Compute number of bytes mod 64 */
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((UINT4)inputLen << 3))
   < ((UINT4)inputLen << 3))
 context->count[1]++;
  context->count[1] += ((UINT4)inputLen >> 29);

  partLen = 64 - index;

  /* Transform as many times as possible.
*/
  if (inputLen >= partLen) {
 MD5_memcpy
   ((POINTER)&context->buffer[index], (POINTER)input, partLen);
 MD5Transform (context->state, context->buffer);

 for (i = partLen; i + 63 < inputLen; i += 64)
   MD5Transform (context->state, &input[i]);

 index = 0;
  }
  else
 i = 0;

  /* Buffer remaining input */
  MD5_memcpy
 ((POINTER)&context->buffer[index], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
void MD5::MD5Final (unsigned char digest[16], MD5_CTX *context)
{
  unsigned char bits[8];
  unsigned int index, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64.
*/
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);

  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)context, 0, sizeof (*context));
}

unsigned char* MD5::ComputeHash(unsigned char* src,int len)
{
   MD5_CTX context;
   static unsigned char digest[16];

   MD5Init(&context);
   MD5Update(&context, src, len);
   MD5Final(digest, &context);

   return digest;
}

/* MD5 basic transformation. Transforms state based on block.
 */
void MD5::MD5Transform (UINT4 state[4],unsigned char block[64])
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */


  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information.

*/
  MD5_memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
void MD5::Encode(unsigned char *output, UINT4 *input, unsigned int len)
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
 output[j] = (unsigned char)(input[i] & 0xff);
 output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
 output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
 output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
void MD5::Decode (UINT4 *output, unsigned char *input, unsigned int len)
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
 output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
   (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
}

/* Note: Replace "for loop" with standard memcpy if possible.
 */

void MD5::MD5_memcpy (POINTER output, POINTER input, unsigned int len)
{
  unsigned int i;

  for (i = 0; i < len; i++)

 output[i] = input[i];
}

/* Note: Replace "for loop" with standard memset if possible.
 */

void MD5::MD5_memset (POINTER output,int value,unsigned int len)
{
	unsigned int i;
	for (i = 0; i < len; i++)
		((char *)output)[i] = (char)value;
}
